packages/core/src/extensions/Blocks/api/customBlocks.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  createInternalBlockSpec,
+  createStronglyTypedTiptapNode,
+  getBlockFromPos,
+  parse,
+  propsToAttributes,
+  wrapInBlockStructure,
+} from "./block";
+import {
+  BlockFromBlockConfig,
+  BlockSchemaWithBlock,
+  PropSchema,
+} from "./blockTypes";
+
+// Defines the config for a custom block. Meant to be used as an argument to
+// `createBlockSpec`, which will create a new block spec from it.
+export type CustomBlockConfig = {
+  type: string;
+  readonly propSchema: PropSchema;
+  containsInlineContent: boolean;
+};
+
+export type BlockConfigFromCustomBlockConfig<T extends CustomBlockConfig> =
+  Omit<T, "containsInlineContent"> & {
+    content: T["containsInlineContent"] extends true ? "inline" : "none";
+  };
+
+export type BlockFromCustomBlockConfig<T extends CustomBlockConfig> =
+  BlockFromBlockConfig<BlockConfigFromCustomBlockConfig<T>>;
+
+export type CustomBlockImplementation<T extends CustomBlockConfig> = {
+  render: (
+    /**
+     * The custom block to render
+     */
+    block: BlockFromBlockConfig<BlockConfigFromCustomBlockConfig<T>>,
+    /**
+     * The BlockNote editor instance
+     * This is typed generically. If you want an editor with your custom schema, you need to
+     * cast it manually, e.g.: `const e = editor as BlockNoteEditor<typeof mySchema>;`
+     */
+    editor: BlockNoteEditor<BlockSchemaWithBlock<T["type"], T["propSchema"]>>
+    // (note) if we want to fix the manual cast, we need to prevent circular references and separate block definition and render implementations
+    // or allow manually passing <BSchema>, but that's not possible without passing the other generics because Typescript doesn't support partial inferred generics
+  ) => {
+    dom: HTMLElement;
+    contentDOM?: HTMLElement;
+    destroy?: () => void;
+  };
+  // Exports block to external HTML. If not defined, the output will be the same
+  // as `render(...).dom`. Used to create clipboard data when pasting outside
+  // BlockNote.
+  // TODO: Maybe can return undefined to ignore when serializing?
+  toExternalHTML?: (
+    block: BlockFromBlockConfig<BlockConfigFromCustomBlockConfig<T>>,
+    editor: BlockNoteEditor<BlockSchemaWithBlock<T["type"], T["propSchema"]>>
+  ) => {
+    dom: HTMLElement;
+    contentDOM?: HTMLElement;
+  };
+};
+
+// A function to create custom block for API consumers
+// we want to hide the tiptap node from API consumers and provide a simpler API surface instead
+export function createBlockSpec<T extends CustomBlockConfig>(
```

Comment:
Do we need to split the config & implementation when creating a custom block (since this is a breaking change)? I understand it helps to mirror the structure of `BlockSpec`, but `implementation`/`blockImplementation` is different between them either way.

Reply:
I think yes. The issue is that I don't think it's easily possible to have 1 object where some properties depend of the value of another property, which is the case that we had with `render(block)` (type of block depends on the rest of the object). At least, I couldn't get it working